moveit_planners/ompl/ompl_interface/src/parameterization/joint_space/constrained_planning_state_space_factory.cpp

```diff
@@ -0,0 +1,58 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2020, KU Leuven
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of KU Leuven nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Jeroen De Maeyer */
+/* Mostly copied from Ioan Sucan's code */
+
+#include <moveit/ompl_interface/parameterization/joint_space/constrained_planning_state_space.h>
+#include <moveit/ompl_interface/parameterization/joint_space/constrained_planning_state_space_factory.h>
+
+ompl_interface::ConstrainedPlanningStateSpaceFactory::ConstrainedPlanningStateSpaceFactory()
+  : ModelBasedStateSpaceFactory()
+{
+  type_ = ConstrainedPlanningStateSpace::PARAMETERIZATION_TYPE;
+}
+
+int ompl_interface::ConstrainedPlanningStateSpaceFactory::canRepresentProblem(
+    const std::string& /*group*/, const moveit_msgs::MotionPlanRequest& /*req*/,
+    const moveit::core::RobotModelConstPtr& /*robot_model*/) const
+{
+  return -2;
+}
```

Comment:
This is related to the rather complicated approach to select state spaces. It is documented in the corresponding header file:
```C++
  /** \brief Return a priority that this planner should be used for this specific planning problem.
   *
   * This state space factory is currently only used if `use_ompl_constrained_state_space` was set to `true` in
   * ompl_planning.yaml. In that case it is the only factory to choose from, so the priority does not matter.
   * It returns a low priority so it will never be choosen when others are available.
   * (The second lowest priority is -1 in the PoseModelStateSpaceFactory.)
   *
   * For more details on this state space selection process, see:
   * https://github.com/JeroenDM/moveit/pull/2
   * **/
```

I will add a comment in the source file that quickly explains it and point to the header file for more info.

Reply:
Something like this:
```C++
  // Return the lowest priority currently existing in the ompl interface.
  // This state space will only be selected if it is the only option to choose from.
  // See header file for more info.
return -2;
```

Reply:
Ah, I'm sorry I didn't understand that.  Thank you for clarifying.